addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/filesystem-role-service.js

```diff
@@ -0,0 +1,553 @@
+      .projection(fields)
+      .get();
+
+    return toFsRoleEntity(dbEntity);
+  }
+
+  /**
+   * This method is similar to the 'find' method but it throws an exception if the filesystem role is not found.
+   *
+   * @param requestContext The standard requestContext
+   * @param accountId The data source account id that the filesystem role belongs to
+   * @param bucket The name of the bucket
+   * @param arn The filesystem role arn
+   * @param fields An array of the attribute names to return, default to all the attributes of the
+   * filesystem role entity.
+   */
+  async mustFind(requestContext, { accountId, bucket, arn, fields = [] }) {
+    const result = await this.find(requestContext, { accountId, bucket, arn, fields });
+    if (!result) throw this.boom.notFound(`Filesystem role with arn "${arn}" does not exist`, true);
+    return result;
+  }
+
+  /**
+   * Call this method to allocate a filesystem role entity for the given study. This method is smart
+   * enough to reuse an existing filesystem role entity if there is one already for the same study.
+   *
+   * Note: this method creates an IAM role resource in an AWS account, if needed.
+   *
+   * @param requestContext The standard requestContext
+   * @param studyEntity The study entity. IMPORTANT, it is expected that this study entity will have
+   * an additional property named 'envPermission', it is an object with the following shape:
+   * { read: true/false, write: true/false }
+   * @param environmentEntity The environment entity that will use the filesystem role
+   */
+  async allocateRole(requestContext, studyEntity = {}, environmentEntity = {}, memberAccountId) {
+    // Allocating a filesystem role is only applicable for bucket with access = 'roles'
+    if (studyEntity.bucketAccess !== 'roles') return undefined;
+
+    // Perform default condition checks to make sure the user is active
+    await this.assertAuthorized(
+      requestContext,
+      { action: 'allocate-fs-role', conditions: [allowIfActive, allowIfAdmin] },
+      studyEntity,
+    );
+
+    const { id: envId } = environmentEntity;
+    const { accountId, id: studyId, bucket, appRoleArn } = studyEntity;
+    const studyResourceId = studyResourceUsageId(studyEntity);
+    const fsUsageSetName = `fs-roles-study-${studyId}`;
+    const [appRoleService, resourceUsageService] = await this.service([
+      'roles-only/applicationRoleService',
+      'resourceUsageService',
+    ]);
+
+    // Primer
+    // We capture the relations between four items: fs role, study, member account and environment.
+    // We keep the relationship information in three database items across two tables.
+    // 1. Study to fs roles, this is a one-to-many relationship (but we need to make it many-to-many in the future
+    //    if we introduce storage gateway). This relationship is captured in the ResourceUsages table.
+    //    Database item => resource: <studyId>, setName: <studyId>, items: [<fs arn>, ...]
```

Comment:
I'm a bit confused by the key structure of ResourceUsages table. Is <studyId> the primary key? Or <studyId, setName> as composite key?

Reply:
It is a composite key (resource, setName).  I wish I can attached a diagram here.  For a description of the tables, you can take a look at the dev quip doc that we have.

The comments in the file system role service are not meant to introduce the resource usage service.  The resource usage service was introduced in a previous PR.  Hopefully the documentation for that service is good enough.  Let me know.